split_at for Array

vbarrielle opened the pull request:
This PR introduces a `split_at`like family of functions for arrays. This is currently work in progress.

Status:
- [x] `axis_split_at`
- [x] `axis_split_at_mut`
- [x] tests

bluss commented:
I think I prefer if we only provide axis split (not outer split, since it's a simple special case, isn't it?)

bluss commented on src/lib.rs:
```diff
@@ -1395,6 +1395,57 @@ impl<A, S, D> ArrayBase<S, D> where S: Data<Elem=A>, D: Dimension
+
+        let mut dim_left = self.dim.clone();
+        dim_left.slice_mut()[axis] = index;
+        let left = ArrayView {
```
You can use `ArrayView::new_` here.

vbarrielle replied:
right, that's less boilerplate

vbarrielle commented:
Indeed it's just a special case, but I was wondering if we didn't want to have some symmetry with the `outer_iter`, `axis_iter` methods?

bluss commented:
The most flexible API is not the easiest to use. Again a view based `self -> (Self, Self)` function may be the easiest.

Consider a function like this:

`fn divide_and_conquer<'a, A>(array: ArrayView<'a, A, Ix>) { }`

The only way to pass on a view with the same lifetime `'a` is the `self` based methods I think. We want to be able to split an array view without tying the split result to the stack frame where the split happens.

This is the same way `.into_iter()` for an array view works, it is lifetime preserving unlike `.iter()`. So if you examine this problem, you find that maybe you want `into_inner_iter()` etc methods for array views too. (Lifetime-preserving iterator constructors). It's all quite complicated.

bluss commented:
Consistency is good, but we need to stem the rising tide of a million methods too. Maybe I shouldn't have added `OuterIter` either.

vbarrielle commented:
Yes I see your point with the lifetimes issue. I wanted to mimick the slice's API but I guess there is some special treatment to slices here.

One (probably ugly) way I found to circumvent that problem in other crates (sprs) was to have additional methods defined only on views which would return views based on the lifetime of the data, not the lifetime of self in the method call. I wish there were some sort of "lifetime specialization" to overcome this issue. Not sure I'm really clear here, my thoughts are a bit blurry and I think I lack the proper vocabulary to express them.

bluss commented:
Yes! Slices are dsts, so they implement methods on the type `[T]`. If you look closely, you'll find lots of traces of this issue where it would be cool if an `ArrayView` could be a custom DST like that.

I think that `ArrayView::axis_split_at(self)` and `ArrayViewMut::axis_split_at(self)` is what we want here. We have to solve view methods not showing up in the docs, though.

vbarrielle commented:
> We have to solve view methods not showing up in the docs, though.

If I remember correctly, I've had a similar issue in sprs, and the solution was to define the impl block on the full type, not the alias type. That was quite some time ago but it might be worth a try.

bluss commented:
Yes, that will work

bluss commented:
The for loop doesn't seem right.

Is this right: consider a 3 x 4 x 5 array (3 pages, 4 rows, 5 columns)

Split on axis = 2, index = 2 should produce two views of shapes
3 x 4 x 2 and 3 x 4 x 3

vbarrielle commented:
I'm not sure I got your point here, that for loop is only supposed to help find the correct pointer offset, but the resulting shapes should be what you said.
However it's getting late here so it's quite probable I'm too sleepy to see the issue.

bluss commented:
I'll review it more carefully. Some pen and paper doodling (that's when you know you're actually thinking..).

vbarrielle commented:
I've switched to the view-by-value pattern and added some tests. I'll probably add a macro to factor out the code once again, and I need to test more complicated cases.

bluss commented:
Awesome

bluss commented on src/lib.rs:
```diff
@@ -879,6 +879,51 @@ impl<'a, A, D> ArrayView<'a, A, D>
+                    *ind = index;
+                }
+            }
+            let offset = self.dim.stride_offset_checked(&self.strides,
```
There's the `stride_offset` method which does this without bounds checking.

bluss replied:
On a closer look, we can just skip the whole for loop, can't we? Using stride * index directly (function `stride_offset` in dimension for example).

bluss commented on src/lib.rs:
```diff
@@ -879,6 +879,51 @@ impl<'a, A, D> ArrayView<'a, A, D>
+                if ax != axis {
+                    *ind = 0;
+                }
+                else {
```
Can you put the `else` on the same line as the `}`? Thanks.

vbarrielle commented:
I've fixed the formatting issue, replaced the for loop by an appropriate call to `dimension::stride_offset`, and added more tests.

I'm no longer sure I should try and factor both impls in a macro, they're quite short now. As you wish :)

bluss commented:
Awesome, this is good. I'll merge the Axis PR too.

bluss commented:
Before I release this, what do you think of renaming it?

`pub fn split_at(self, axis: Axis, index: Ix)`

versus

`pub fn split_at(self, axis: Axis, index: Ix)`

I bring this up since `split_at` seems nice an unambiguous too.

vbarrielle commented:
Yes `split_at` seems nice and unambiguous, I don't think we could ever have a split_at taht does not require an axis argument.

bluss commented:
Done.
